src/core/experimental/memory/memcpy.d

```diff
@@ -0,0 +1,391 @@
+        for (int i = 0; i < m.length; i++)
+        {
+            m[i] = cast(ubyte) i;
+        }
+    }
+    pragma(inline, false)
+    void verifyStaticType(T)(const T *a, const T *b)
+    {
+        const ubyte* aa = (cast(const ubyte*) a);
+        const ubyte* bb = (cast(const ubyte*) b);
+        // NOTE(stefanos): `real` is an exceptional case,
+        // in that it behaves differently across compilers
+        // because it's not a power of 2 (its size is 10 for x86)
+        // and thus padding is added (to reach 16). But, the padding bytes
+        // are not considered (by the compiler) in a move (for instance).
+        // So, Dmemcpy, for static types, is *dst = *src. And the compiler
+        // might output `fld` followed by `fstp` instruction. Those intructions
+        // operate on extended floating point values (whose size is 10). And so,
+        // the padding bytes are not copied to dest.
+        static if (is(T == real))
+        {
+            enum n = 10;
+        }
+        else
+        {
+            enum n = T.sizeof;
+        }
+        for (size_t i = 0; i < n; i++)
+        {
+            assert(aa[i] == bb[i]);
+        }
+    }
+    pragma(inline, false)
+    void testStaticType(T)()
+    {
+        T d, s;
+        initStatic!(T)(&d);
+        initStatic!(T)(&s);
+        memcpy(d, s);
+        verifyStaticType(&d, &s);
+    }
+    pragma(inline, false)
+    void init(T)(ref T[] v)
+    {
+        for (int i = 0; i < v.length; i++)
+        {
+            v[i] = cast(ubyte) i;
+        }
+    }
+    pragma(inline, false)
+    void verifyArray(size_t j, const ref ubyte[] a, const ref ubyte[80000] b)
+    {
+        //assert(a.length == b.length);
+        for (int i = 0; i < a.length; i++)
+        {
+            assert(a[i] == b[i]);
+        }
+    }
+    pragma(inline, false)
+    void testDynamicArray(size_t n)()
+    {
+        ubyte[80000] buf1;
+        ubyte[80000] buf2;
+        enum alignments = 32;
+        foreach (i; 0..alignments)
+        {
+            ubyte[] p = buf1[i..i+n];
+            ubyte[] q;
+            // Relatively aligned
+            q = buf2[0..n];
+            // Use a copy for the cases of overlap.
+            ubyte[80000] copy;
+            pragma(inline, false);
+            init(q);
+            pragma(inline, false);
+            init(p);
+            for (size_t k = 0; k != p.length; ++k)
+            {
+                copy[k] = p[k];
+            }
+            pragma(inline, false);
+            memcpy(q, p);
+            pragma(inline, false);
+            verifyArray(i, q, copy);
+        }
+    }
+}
+
+unittest
+{
+    tests();
+}
+
+/* Implementation
+ */
+
+import core.experimental.memory.simd;
+
+/*
+ * Dynamic implementation
+ * NOTE: Dmemcpy assumes _no_ overlap
+ *
+ */
+static if (useSIMD)
+{
+
+import core.simd : float4;
+
+/* Handle dynamic sizes. `d` and `s` must not overlap.
+ */
+void Dmemcpy(void* d, const(void)* s, size_t n)
```

Comment:
`private`

Reply:
That is public because of the description of this PR:

> This is a D replacement for the C Standard Library memcpy. It provides 2 interfaces:

> The "D" one, which takes, one of ref, dynamic arrays or static arrays.
> The "C" one, via the name Dmemcpy (for disambiguation with the libc), which takes the classic: `(void*, void*, size_t)`.

So, because the idea was to provide the C interface as well. Is this desired?
Otherwise, yes it should be `private`.

Reply:
Isn't that the C version then? I think all bases should be covered with static arrays and dynamic arrays, you can always slice a pointer to create a dynamic array from it and a length..

Reply:
Yes, it is the C version. It is named _D_`memcpy` for disambiguation with the libc `memcpy`.
We changed the `Dmemcpy` name to `memcpy` for the D interface, which provides 3 overloads, the ones you can see on top of the file here, named `memcpy`. I guess this was done with the purpose that they are overloads for the libc `memcpy`. They provide handling for `ref`s, static arrays and dynamic arrays.
But I thought it would be good to also provide a C interface but for the D version (aka the druntime version) named `Dmemcpy`, that is different from the libc.

Reply:
OK.

Reply:
So, LDC/GDC would never use this function right... they're re-route to intrinsics?

Reply:
They will use the SIMD version. They will probably use their intrinsics for when SIMD is not available.

Reply:
They should use their intrinsics unconditionally. There's no way we should take that knowledge away from the optimiser.

Reply:
That is something important that I have thought and that Johan Engelen pointed out as well: https://github.com/dlang/druntime/pull/2662#issuecomment-517742489

For one, to the best of my knowledge (meaning, by inspecting the assembly), ldc and gdc, when they reroute, they re-route to the c standard library. They just, as you said, know that it is a memory copy, with all the implications. But this is what we try to avoid here (meaning the libc).
I don't know what ldc and gdc do if libc is not available. Do they just use our function or they have some other intrinsic? If they have their intrinsic, what code is it? Should we use that or ours but somehow give the compiler info that this is a memory copy ?
And I don't know of a way to give ldc and gdc such info.

Reply:
Exactly; all the things he said. For my money, re-implementing memcpy and memset for LDC/GDC is a very bad idea. Just don't do it. Even if you test some cases and we don't see a disadvantage today, it might appear tomorrow, next thursday, in may 2022...
You should assume this whole thing only applies only to DMD, and then you need to carefully consider that compiler's optimisation capabilities.

Reply:
As I said here: https://github.com/dlang/druntime/pull/2687#discussion_r310724318
The focus was initially DMD. Then that was considered a bad idea so we switched to LDC / GDC.
For me, depending on the goal, targetting DMD is not a very good idea. Unless we either want to write ASM or not reach libc.

Now, that seems a good idea if there's a problem with intrinsics if libc is not available. It's also a good idea if one plans to completely get rid of libc (this is not planned). But it also seems good if one wants to have their own version but somehow can instruct the compiler about what the function does.

Reply:
I don't understand of follow any of those propositions, but I'll assume that the discussion has been had elsewhere and makes sense.
I don't know what your goals are, I was just browsing the PR. At face value though, the motive feels weird... is there a reference to the goals and initial discussion? You mentioned that the goals changed, I'd kinda like to see how and why?

Reply:
> I don't understand of follow any of those propositions, but I'll assume that the discussion has been had elsewhere and makes sense.

Could you be more specific? You seem to like DMD as a target, in which case I think at least one thing is true. Targetting DMD is a good idea if we either want to write ASM or we don't care for parity with libc.

Other than that, this specific discussion has not been done elsewhere and I too try to understand more and more things as this project goes forward and the oppositions are more and more. Like yours, Johan's, Iain's, Andrei's initially etc. Thank you for your time, I mean it.
But just have in your mind that:
I don't have a mentor or anyone interested in this project for that matter. So, I face the oppositions myself for a lot of things that I don't know and / or have experience with. Like the compiler intrinsics. And most of the oppositions are not about something I _wrote_, like "hey, why is this mul there?". Yes, we can discuss this, I'm supposed to have knowledge about the code I've written. But for the project _goals_.. I don't know, I don't think it's the best idea to leave to a student to justify why the project is needed. And btw, I'm not too very sure about this project.

Anyway, the initial goals were these: https://forum.dlang.org/thread/izaufklyvmktnwsrmhci@forum.dlang.org

Which basically is DMD and close performance to libc. To have that, one has to write ASM, so most of the code was inline ASM.
You can see some very early oppositions I had about the project goals very early on: https://forum.dlang.org/post/juipzgghcwcengyackgt@forum.dlang.org
And that was with much less knowledge about what can go wrong.

Then, after the project was left without a mentor, the goals changed to:
https://forum.dlang.org/post/qbicnduehcmxeaqwervf@forum.dlang.org

And _all_ the code was re-written to D.

Reply:
Actaully... I don't really care much at all about DMD; it's mostly a toy. But what I'm saying is that DMD could absolutely *benefit* from good implementations of these functions, whereas LDC/GDC will likely rather suffer.

> I don't have a mentor or anyone interested in this project for that matter. So, I face the oppositions myself for a lot of things that I don't know and / or have experience with.

You're diving into some really complex and heavy shit, so I give you credit for working on this challenging stuff!

> Anyway, the initial goals were these: [...]

Thanks, I didn't actually know the details of your project ;)

Okay, I think I can see something about your process here. Sorry for jumping in and messing the waters so late. I was just reviewing some PR's and saw what looked like very interesting PR's... but as you have determined, also very complex ;)

Who was the mentor initially? Have you had any offers for a replacement?

To be clear; I'm not against your current work. The goals of your project are interesting and useful (although libc independence is not interesting to me personally, there's a libc everywhere I've ever worked), but if you want to do it, then it's important to understand the context it must be plugged into.
The goals seem to have changed, which is never fun, but at very least you're learning a whole lot about a very complicated ecosystem. Progress here was never going to be easy ;)

The intrinsics do call out to library functions when the compilers determine that's what they want to do... so supplying a source for those symbols other than libc may be good for D to have, but producing optimal code is unlikely, and frankly, not worth the time invested. You should probably define initial success as a 'compatibility grade' implementation, which would be a dignified outcome for users that specifically want a stand-alone environment. No need to waste a heap of time benchmarking and micro-tuning on those terms.

If you take it from that perspective, then as-portable-as-possible implementations with no asm or any nonsense makes sense, and what you could look into, is defining these functions in druntime but specified with weak-linkage, then the libc functions will override them if the user links a libc.

But in addition to that, I think it's very important to review the GCC and LDC intrinsics, understand how they are used by the optimiser, and how the compilers will emit library calls when they choose to. Do they call the standard memcpy/set functions in the libc, or are there other implementations integrated into the toolchain's runtime libraries designed to work together with the intrinsics?
It's probably the case that they just call out to libc, but I think it's important to understand how and when that works when considering your implementations here.

I think @ibuclaw and @JohanEngelen will have much better advice than me on this whole matter; they are experts on this.

Reply:
> Actaully... I don't really care much at all about DMD; it's mostly a toy. But what I'm saying is that DMD could absolutely benefit from good implementations of these functions, whereas LDC/GDC will likely rather suffer.

Oh, I agree, me neither. From a personal point of view, it was really fun optimizing for DMD exactly because it doesn't have the GCC / LLVM optimizer. But for D, I don't think it's worth. So yes, I see that DMD could definitely benefit from this, but I don't think we should invest time.

> You're diving into some really complex and heavy shit, so I give you credit for working on this challenging stuff!

Thank you! You know, unfortunately the challenging stuff turned out not to be what I expected / told it would be (aka reaching libc in performance) rather turned out to be things about I don't have experience or knowledge (aka justifying the goals, specifics about compiler intrinsics etc.).

> Thanks, I didn't actually know the details of your project ;)
>
> Okay, I think I can see something about your process here. Sorry for jumping in and messing the waters so late. I was just reviewing some PR's and saw what looked like very interesting PR's... but as you have determined, also very complex ;)

No thank you for that. And all of you. You know these PRs might never be merged, but at least I'll know why and I will have learned something.

> Who was the mentor initially?

Mostly Mike Franklin.

> Have you had any offers for a replacement?

No. Although I thank @thewilsonator for jumping in and helping me move these PRs forward (and for other things, but anyway).

> To be clear; I'm not against your current work. The goals of your project are interesting and useful (although libc independence is not interesting to me personally, there's a libc everywhere I've ever worked), but if you want to do it, then it's important to understand the context it must be plugged into.
The goals seem to have changed, which is never fun, but at very least you're learning a whole lot about a very complicated ecosystem. Progress here was never going to be easy ;)

Yes, indeed! Well, again, the tough part is that the problems are things I don't know. And things that require experience. I was unsure for the goals of this project in the first 5 days or so. Then, even when these goals changed, there were unexpected problems that I just didn't know of. Problems with the very _goals_ not the implementation.

> The intrinsics do call out to library functions when the compilers determine that's what they want to do...

Yes that seems reasonable.

> so supplying a source for those symbols other than libc may be good for D to have, but producing optimal code is unlikely, and frankly, not worth the time invested.
You should probably define initial success as a 'compatibility grade' implementation, which would be a dignified outcome for users that specifically want a stand-alone environment. No need to waste a heap of time benchmarking and micro-tuning on those terms.

I might have missed this. You mean, an implementation but not e.g. with a constraint to be close to libc. Just an implementation that is D, it is general and it works.

> If you take it from that perspective, then as-portable-as-possible implementations with no asm or any nonsense makes sense, and what you could look into, is defining these functions in druntime but specified with weak-linkage, then the libc functions will override them if the user links a libc.

Oh, yeah, I see you mention a similar thing. Weak-linkage is a term? I understand the meaning from the context, but I've never heard of it.

> But in addition to that, I think it's very important to review the GCC and LDC intrinsics, understand how they are used by the optimiser, and how the compilers will emit library calls when they choose to. Do they call the standard memcpy/set functions in the libc, or are there other implementations integrated into the toolchain's runtime libraries designed to work together with the intrinsics? Is that decision back-end specific, could it change for an odd platform?
It's probably the case that they always just call out to libc, but I think it's important to understand how and when that works when considering your implementations here.

Intrinsics had been mentioned initially. But the focus was DMD then, so nobody really cared.
Even when it moved to LDC / GDC, I didn't know that the optimizers _know_ about the behavior of e.g. memcpy (i.e. after the call, this will have a copy of this). Then Johan happened :p. Thanks Johan.
Obviously, with that information, and considering that anyone who wants performance in x86_64 (this project's performance is focused on x86_64) will probably have libc, then this project becomes kind of irrelevant.

> I think @ibuclaw and @JohanEngelen will have much better advice than me on this whole matter; they are experts on this.

Their advice is welcome!

Reply:
> this project becomes kind of irrelevant.

Don't forget web assembly and similar targets.

Reply:
The thing is that the whole code and time invested is optimizing for x86 (and x86_64). For WASM and other platforms, the compiler will branch to the naive versions, which are like _really_ naive. Meaning, a musl-like implementation (which doesn't have SIMD) would be more appropriate if such platforms were the main focus.

Reply:
> Weak-linkage is a term? I understand the meaning from the context, but I've never heard of it.

Yes, see e.g. [ldc.attributes.weak](https://wiki.dlang.org/LDC-specific_language_changes#.40.28ldc.attributes.weak.29), but note that D sadly doesn't have a standard attribute for it and that it's not supported on all platforms.

Reply:
Ah, good to know!

Reply:
> it's not supported on all platforms.

Where is it not supported?

D just doesn't have a standard for it... but that's very possible to fix.

Reply:
> Where is it not supported?

AFAIK MSVC (or whatever object format that Windows uses, COFF/PE?) doesn't support weak linkage, but I might be wrong.

> D just doesn't have a standard for it... but that's very possible to fix.

I agree, would be great to have this.

Reply:
> but I might be wrong.

Yeah. I'm pretty sure weak linkage is available everywhere, and it might be particularly useful in this project.

I wonder if there's any reason that we don't have a weak attribute already? I think I've asked for it in the past and it was denied because reasons... I wonder what they were :/

Reply:
Function should still be private.

Reply:
> Yeah. I'm pretty sure weak linkage is available everywhere

The accepted answer and the most up-voted answer on StackOverflow[1] suggest that at least the semantics are different with MSVC. Other compilers/linkers might implement workarounds though, I don't have access to a Windows machine to test on.

[1] https://stackoverflow.com/questions/2290587/gcc-style-weak-linking-in-visual-studio?answertab=votes#tab-top